README.md
# Stations Set Train Colour

I couldn't find any mod doing exactly this, so I wrote my own

This mod allows you to set the colour of a train when it arrives (and waits) at a station. Note - nothing will happen unless the train actually stops.

## Setting the colour

Connect a circuit network to a station and set the ![sstc-set-colour](https://github.com/ccooke/factorio-StationsSetTrainColour/blob/main/icons/sstc-set-color.png?raw=true) signal.
When a train stops, the `R`, `G`, `B` and `A` signals will be read and used to set the colour and transparency. They default to 0,0,0,128.

## Resetting the colour

Connect a circuit network to a station and set the ![sstc-reset-colour](https://github.com/ccooke/factorio-StationsSetTrainColour/blob/main/icons/sstc-reset-color.png?raw=true) signal.
When a train stops, the train's colour will be reset to the default.

You can also set the module option `Default Reset`, which causes all stations to reset colour unless the ![sstc-set-colour](https://github.com/ccooke/factorio-StationsSetTrainColour/blob/main/icons/sstc-set-color.png?raw=true) signal is set.

## Precedence

If the ![sstc-reset-colour](https://github.com/ccooke/factorio-StationsSetTrainColour/blob/main/icons/sstc-reset-color.png?raw=true) signal is set, the station will *always* reset the train's colour to default.

## Module Options

* `Default Reset` - Reset train colours by default
* Default `Red`, `Green`, `Blue` and `Alpha` - The default colour for trains. Initially set to the vanilla standard of [234,17,0,128]
* `Debug` - Generate debug messages
* `Red`, `Green`, `Blue` and `Alpha` signals - The names of the signals to use for colour. *Do not change these unless you know what you are doing*


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.0.5
Date: 2021-01-09 10:49:03+00:00
  Bugfixes:
    - Fixed colour error even when no actions were to be taken
    - Fixed colour error preventing resets
---------------------------------------------------------------------------------------------------
Version: 0.0.4
Date: 2020-01-09 03:24:50+00:00
  Info:
    - Added a better-than-nothing thumbnail
---------------------------------------------------------------------------------------------------
Version: 0.0.3
Date: 2020-01-09 02:54:01+00:00
  Bugfixes:
    - Fixed 'Default Reset' not working when stations had no signals
---------------------------------------------------------------------------------------------------
Version: 0.0.2
Date: 2020-01-09 02:11:28+00:00
  Feature:
    - Module settings added
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 2020-01-08 19:23:15+00:00
  Info:
    - Initial release
  Feature:
    - Set train colour from signals
  Locale:
    - English only


settings.lua
data:extend({
  {
    type = "string-setting",
    name = "sstn-rgb-red-signal",
    order = "da",
    setting_type = "runtime-global",
    default_value = "signal-R",
  },
  {
    type = "string-setting",
    name = "sstn-rgb-green-signal",
    order = "db",
    setting_type = "runtime-global",
    default_value = "signal-G",
  },
  {
    type = "string-setting",
    name = "sstn-rgb-blue-signal",
    order = "dc",
    setting_type = "runtime-global",
    default_value = "signal-B",
  },
  {
    type = "string-setting",
    name = "sstn-rgb-alpha-signal",
    order = "dd",
    setting_type = "runtime-global",
    default_value = "signal-A",
  },
  {
    type = "bool-setting",
    name = "sstn-default-reset",
    order = "aa",
    setting_type = "runtime-global",
    default_value = false
  },
  {
    type = "bool-setting",
    name = "sstn-debug",
    order = "ab",
    setting_type = "runtime-global",
    default_value = false
  },
  {
    type = "int-setting",
    name = "sstn-default-rgb-red",
    order = "ba",
    setting_type = "runtime-global",
    default_value = 234,
    minimum_value = 0,
    maximum_value = 255
  },
  {
    type = "int-setting",
    name = "sstn-default-rgb-green",
    order = "bb",
    setting_type = "runtime-global",
    default_value = 17,
    minimum_value = 0,
    maximum_value = 255
  },
  {
    type = "int-setting",
    name = "sstn-default-rgb-blue",
    order = "bc",
    setting_type = "runtime-global",
    default_value = 0,
    minimum_value = 0,
    maximum_value = 255
  },
  {
    type = "int-setting",
    name = "sstn-default-rgb-alpha",
    order = "bd",
    setting_type = "runtime-global",
    default_value = 128,
    minimum_value = 0,
    maximum_value = 255
  }
})



control.lua
--Detect the state

-- Define the signals we will use
RGB = {
  [settings.global["sstn-rgb-red-signal"].value] = "r",
  [settings.global["sstn-rgb-green-signal"].value] = "g",
  [settings.global["sstn-rgb-blue-signal"].value] = "b",
  [settings.global["sstn-rgb-alpha-signal"].value] = "a"
}

-- The icon used for our alert signal
ERROR = { ["type"] = "virtual", ["name"] = "sstc-color-error" }

-- flags used to select actions
MODES = {
  do_nothing = 0,
  set_colour = 1,
  reset_colour = 2
}

-- What signals activate modes
CONTROL_SIGNALS = {
  ["sstc-set-color"] = MODES.set_colour,
  ["sstc-reset-color"] = MODES.reset_colour
}

-- Strings for debugging
SET = 'Set (by signal)'
RESET = 'Reset (by signal)'
DEFAULT_RESET = 'Reset'

function default_train_colour()
  return({
    r=settings.global["sstn-default-rgb-red"].value,
    g=settings.global["sstn-default-rgb-green"].value,
    b=settings.global["sstn-default-rgb-blue"].value,
    a=settings.global["sstn-default-rgb-alpha"].value
  })
end

function colour_to_string(colour)
  return "(R="..colour.r..",G="..colour.g..",B="..colour.b..",A="..colour.a..")"
end

function set_train_colour(train, colour, action)
  if settings.global['sstn-debug'].value then
    game.print("Train "..train.id.." waiting at station "..train.station.backer_name.." - "..action.." to "..colour_to_string(colour))
  end
  for _,locomotive in pairs(train.locomotives['front_movers']) do
    locomotive.color = colour
  end
  for _,locomotive in pairs(train.locomotives['back_movers']) do
    locomotive.color = colour
  end
end

function player_alert(entity, icon, message, force)
  force = force or (entity and entity.force)
  if not force or not force.valid then
    return
  end
  for _, player in pairs(force.players) do
    player.add_custom_alert(entity, icon, message, true)
    player.play_sound{path="utility/cannot_build"}
  end
end

script.on_event(defines.events.on_train_changed_state,
  function (event)
    local train = event.train
    if train.state == defines.train_state.wait_station and train.station ~= nil then
      local station = train.station
      local signals = station.get_merged_signals()
      local out = serpent.line(signals)

      local rgb = { r=0, g=0, b=0, a=128 }

      if signals == nil then
        signals = {}
      end

      local mode = MODES.do_nothing
      local colour_error = false

      for _,s in pairs(signals) do
        if RGB[s.signal.name] then
          if s.count > 255 or s.count < 0 then
            colour_error = true
          else
            rgb[RGB[s.signal.name]] = s.count
          end
        elseif CONTROL_SIGNALS[s.signal.name] and s.count > 0 then
          mode = bit32.bor(mode, CONTROL_SIGNALS[s.signal.name])
        end
      end

      if settings.global['sstn-debug'].value then
        game.print("Train "..train.id.." waiting at station "..train.station.backer_name.." - MODE is "..mode.." --- reset=".. bit32.band(mode,MODES.reset_colour) .. ", set="..bit32.band(mode,MODES.set_colour))
      end
      if bit32.band(mode, MODES.reset_colour) > 0 then
        set_train_colour(train, default_train_colour(), RESET)
      elseif bit32.band(mode, MODES.set_colour) > 0 then
        if colour_error then
          player_alert(station, ERROR, "RGB values out of range (0-255) at station "..station.backer_name, station.force)
        else
          set_train_colour(train, rgb, SET)
        end
      elseif settings.global['sstn-default-reset'].value then
        set_train_colour(train, default_train_colour(), DEFAULT_RESET)
      end
    end
  end
)

